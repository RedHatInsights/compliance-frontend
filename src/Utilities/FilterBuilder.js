const buildBaseFilter = (state) => {
    const { policyId } = state;
    const name = state.activeFilters ? state.activeFilters.name : [];

    let additionalFilter = [];

    if (policyId && policyId.length > 0) {
        additionalFilter.push(`profile_id = ${policyId}`);
    }

    if (name && name.length > 0) {
        additionalFilter.push(`name ~ ${name}`);
    }

    return [additionalFilter.join(' and ')].filter((f) => (f.length > 0));
};

export const buildFilterString = (state) => {
    const additionalFilter = buildBaseFilter(state);
    const { compliant: complianceStates, compliancescore: complianceScores } = (state.activeFilters || {});
    const compliant = complianceStates ? complianceStates.map((compliant) =>
        `compliant = ${compliant}`
    ) : [];
    const complianceScore = complianceScores ? complianceScores.map((scoreRange) => {
        scoreRange = scoreRange.split('-');
        return `compliance_score >= ${scoreRange[0]} and compliance_score <= ${scoreRange[1]}`;
    }) : [];
    const filters = [
        additionalFilter,
        compliant,
        complianceScore
    ].filter((f) => (f.length > 0));
    const moreThanTwo = filters.map((f) => (f.length)).filter((fl) => (fl > 0)).length >= 2;

    return filters.map((fs) => (fs.join(' or '))).join(moreThanTwo ? ' and ' : '');
};
