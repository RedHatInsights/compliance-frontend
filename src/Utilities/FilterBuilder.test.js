import { buildFilterString } from './FilterBuilder';
import { pickBy } from 'lodash';

const filterKeys = (obj, keys) => (
    pickBy(obj, (_, key) => (keys.includes(key)))
);

describe('buildFilterString', () => {
    const exampleState = {
        policyId: 'exmpleID-1',
        activeFilters: {
            name: 'Name',
            compliant: [true],
            compliancescore: ['0-49', '50-69']
        }
    };

    it('returns a filterstring', () => {
        expect(buildFilterString(exampleState)).toMatchSnapshot();
    });

    it('returns a base filter for a policy id', () => {
        const testExampleState = filterKeys(exampleState, ['policyId']);
        expect(buildFilterString(testExampleState)).toMatchSnapshot();
    });

    describe('compliance state and score filter', () => {
        it('returns a base filter for name when searching', () => {
            const testExampleState = {
                activeFilters: {
                    name: 'Name'
                }
            };
            expect(buildFilterString(testExampleState)).toMatchSnapshot();
        });

        it('returns a filter for complianceStates', () => {
            const testExampleState = {
                activeFilters: {
                    compliant: [true]
                }
            };
            expect(buildFilterString(testExampleState)).toMatchSnapshot();
        });

        it('returns a filter for complianceStates', () => {
            const testExampleState = {
                activeFilters: {
                    compliancescore: ['0-49', '50-69']
                }
            };
            expect(buildFilterString(testExampleState)).toMatchSnapshot();
        });
    });
});
