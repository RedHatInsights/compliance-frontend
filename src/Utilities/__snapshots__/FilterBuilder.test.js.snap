// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildFilterString compliance state and score filter returns a base filter for name when searching 1`] = `"name ~ Name"`;

exports[`buildFilterString compliance state and score filter returns a filter for complianceStates 1`] = `"compliant = true"`;

exports[`buildFilterString compliance state and score filter returns a filter for complianceStates 2`] = `"compliance_score >= 0 and compliance_score <= 49 or compliance_score >= 50 and compliance_score <= 69"`;

exports[`buildFilterString returns a base filter for a policy id 1`] = `"profile_id = exmpleID-1"`;

exports[`buildFilterString returns a filterstring 1`] = `"profile_id = exmpleID-1 and name ~ Name and compliant = true and compliance_score >= 0 and compliance_score <= 49 or compliance_score >= 50 and compliance_score <= 69"`;
